Replace cp_install.sh and cp_init.sh by cpackget

chaws opened the pull request:
Fix https://github.com/Open-CMSIS-Pack/devtools/issues/46

Hi, this is a patch to replace both cp_init.sh and cp_install.sh by cpackget.

I talked to Daniel Brondani last week when he pointed me out to some files that needed changes. I'm pushing this WIP PR and hopefully will split all the changes in a better format in a couple of days.

Here's my current progress:
1. Changed cbuild.sh to use cpackget instead of cp_install. (It seems like cp_init is only used in testing and shipped in the installer)
2. Merging CPInitTests.cpp and CPInstallTests.cpp into CPackGetTests.cpp, where pack root init and pack installation tests happen
2. Changing docs in CmdLineBuild.txt, removed the cp_init section completely
3. Changing create_installer.sh to add a new bin folder, where cpackget binaries will be added

I have two questions:
1. Regarding configuring the installer: I added cpackget (linux) and cpackget.exe, but how to do that for Mac?
2. In docs, there was a table showing off a lot of possible errors messages coming out of cp_init.sh and cp_install.sh. Is this really necessary? I'm asking because cpackget does have lots of error messages. In the docs I wrote that most of the error messages are self explanatory, but the user can always increase verbosity (something that didn't previously existed). Hope that it is OK.

brondani commented:
@chaws Charles, instead of committing binaries into the repository, you should download them from the cpackget releases and rename them according to the platform (.exe, .lin and .mac) during the installer creation.

For example in the `create_installer.sh` script you could add something like this:
```
mkdir temp

curl -L https://github.com/Open-CMSIS-Pack/cpackget/releases/download/v0.1.3/cpackget_0.1.3_linux_amd64.tar.gz -o temp/linux.tar.gz
tar -xzf temp/linux.tar.gz -C temp
cp temp/cpackget_0.1.3_linux_amd64/cpackget bin/cpackget.lin

curl -L https://github.com/Open-CMSIS-Pack/cpackget/releases/download/v0.1.3/cpackget_0.1.3_darwin_amd64.tar.gz -o temp/darwin.tar.gz
tar -xzf temp/darwin.tar.gz -C temp
cp temp/cpackget_0.1.3_darwin_amd64/cpackget bin/cpackget.mac

curl -L https://github.com/Open-CMSIS-Pack/cpackget/releases/download/v0.1.3/cpackget_0.1.3_windows_amd64.zip -o temp/windows.zip
unzip temp/windows.zip -d temp
cp temp/cpackget_0.1.3_windows_amd64/cpackget.exe bin/cpackget.exe

rm -rf temp
```

Concerning documenting error messages, I believe it is not important for this PR but you should keep track of it in the backlog.

chaws commented:
Thanks for the answer, @brondani ! I placed the binaries there for testing purposes. Once my local cpackget binary is passing the integration tests, I'll make a release of it and download as your example.

chaws commented:
Hi @brondani , I fixed tests regarding cpackget, now I'm getting a failure in [DebPkgTests.Extract_Completion](https://github.com/Open-CMSIS-Pack/devtools/blob/main/tools/buildmgr/test/integrationtests/src/DebPkgTests.cpp#L49). I think this checks if the installed debian package contains the required files.

How to proceed here? I have added the cpackget file in the [make_deb.sh](https://github.com/Open-CMSIS-Pack/devtools/pull/50/files#diff-a3d1767228812b50d304fd6c11b24aaa9a113f8773c14a2a1b3ff38488aaad5fR98) but still no luck there

brondani commented:
@chaws From the test report: `./usr/bin/cpackget not found`
It happens because cpackget in this directory is just a symbolic link. To fix it the following line should be removed:
https://github.com/chaws/devtools/blob/cpackget-integration/tools/buildmgr/test/integrationtests/src/DebPkgTests.cpp#L55
The test case could be extended by asserting with `fs::is_symlink` that `./usr/bin/cpackget` and `./usr/bin/cbuildgen` are symbolic links.

chaws commented:
Thanks @brondani !

I opted for extending the test scope to check if the file is a symlink on both `./usr/bin/cpackget` and `./usr/bin/cbuildgen`. Tests are all green. I believe the patch is good now, let me split it into smaller commits then remove "DRAFT" status.

brondani commented on tools/buildmgr/cbuildgen/installer/create_installer.sh:
```diff
@@ -66,10 +66,18 @@ installer_name=cbuild_install.sh
 mkdir -p ${distdir}/bin
 mkdir -p ${distdir}/etc
 
+cp -R ../bin/. ${distdir}/bin
```
I believe this line is not needed (`../bin` does not exist).

chaws replied:
Ack

brondani commented on tools/buildmgr/docs/doxygen/Build/src/CmdLineBuild.txt:
```diff
@@ -735,27 +651,29 @@ M650: Command completed successfully.
+\section cpackget_invocation Invocation
 
-\b cp_install is called from the \b Bash command line with the following syntax:
+\b cpackinstall is called from the \b Bash command line with the following syntax:
```
Does `cpackinstall` have a meaning? Shouldn't it be `cpackget` instead?

brondani commented on tools/buildmgr/cbuildgen/scripts/cbuild.sh:
```diff
@@ -323,7 +323,7 @@ fi
   then
   dos2unix "${abs_intdir}${project}.cpinstall" >&${out} 2>&${err}
-  cp_install.sh "${abs_intdir}${project}.cpinstall" >&${out} 2>&${err}
+  cpackget pack add --agree-embedded-license --packs-list-filename "${abs_intdir}${project}.cpinstall" >&${out} 2>&${err}
```
I would suggest to add the `-v` option for printing info messages, otherwise the cpackget call is completely silent.
```suggestion
  cpackget pack add -v --agree-embedded-license --packs-list-filename "${abs_intdir}${project}.cpinstall" >&${out} 2>&${err}
```

chaws commented:
Thanks for reviewing @brondani ! I have pushed an update that addressed your comments inline. 

Just FYI: Joachim may have talked to you that I'll be working on cpackget so that the file with pack list can made of pack ids instead of pack URLs. I opted for finishing up this PR before the work on cpackget, so next time there's a new release of it I'll just send a PR to devtools changing release version of cpackget.

brondani commented:
Thanks @chaws, it looks good to me!

@soumeh01 can I ask you to review the changes in tests?

soumeh01 reviewed (approved):
LGTM 👍
